File: backends/dotnet/shared/Implementations/Caching/Distributed/DistributedCache.Redis/Handlers/D/Remove.cs
// -----------------------------------------------------------------------
// <copyright file="Remove.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Shared.DistributedCache.Redis.Handlers.D;

using D2.Shared.Handler;
using D2.Shared.Result;
using Microsoft.Extensions.Logging;
using StackExchange.Redis;
using H = D2.Shared.Interfaces.Caching.Distributed.Handlers.D.IDelete.IRemoveHandler;
using I = D2.Shared.Interfaces.Caching.Distributed.Handlers.D.IDelete.RemoveInput;
using O = D2.Shared.Interfaces.Caching.Distributed.Handlers.D.IDelete.RemoveOutput;

/// <summary>
/// Handler for removing a value from the Redis distributed cache.
/// </summary>
public partial class Remove : BaseHandler<H, I, O>, H
{
    private readonly IConnectionMultiplexer r_redis;

    /// <summary>
    /// Initializes a new instance of the <see cref="Remove"/> class.
    /// </summary>
    ///
    /// <param name="redis">
    /// The Redis connection multiplexer.
    /// </param>
    /// <param name="context">
    /// The handler context.
    /// </param>
    public Remove(
        IConnectionMultiplexer redis,
        IHandlerContext context)
        : base(context)
    {
        r_redis = redis;
    }

    /// <inheritdoc/>
    protected override async ValueTask<D2Result<O?>> ExecuteAsync(
        I input,
        CancellationToken ct = default)
    {
        try
        {
            // Connect to Redis and remove the key.
            var db = r_redis.GetDatabase();
            await db.KeyDeleteAsync(input.Key);

            // Return success.
            return D2Result<O?>.Ok(
                new O());
        }
        catch (RedisException ex)
        {
            LogRemoveFailed(Context.Logger, ex, input.Key, TraceId);

            return D2Result<O?>.ServiceUnavailable(traceId: TraceId);
        }

        // Let the base handler catch any other exceptions.
    }

    /// <summary>
    /// Logs that a Redis exception occurred while removing a key.
    /// </summary>
    [LoggerMessage(EventId = 1, Level = LogLevel.Error, Message = "RedisException occurred while removing key '{Key}'. TraceId: {TraceId}")]
    private static partial void LogRemoveFailed(ILogger logger, Exception ex, string key, string? traceId);
}

File: backends/node/shared/implementations/caching/distributed/redis/src/handlers/d/remove.ts
import type Redis from "ioredis";
import { BaseHandler, type IHandlerContext } from "@d2/handler";
import { D2Result } from "@d2/result";
import { redisErrorResult } from "../../redis-error-result.js";
import type { DistributedCache } from "@d2/interfaces";

type Input = DistributedCache.RemoveInput;
type Output = DistributedCache.RemoveOutput;

export class Remove extends BaseHandler<Input, Output> implements DistributedCache.IRemoveHandler {
  private readonly redis: Redis;

  constructor(redis: Redis, context: IHandlerContext) {
    super(context);
    this.redis = redis;
  }

  protected async executeAsync(input: Input): Promise<D2Result<Output | undefined>> {
    try {
      await this.redis.del(input.key);
      return D2Result.ok({ data: {} });
    } catch (err: unknown) {
      return redisErrorResult(err);
    }
  }
}
